README.md
# node-clang-rest

Node based JSON/HTTP proxy to the Clang's SOAP webservice API.

This is a restful webservice that proxies HTTP requests to Clang's SOAP API, passing along any parameters in the query string. All requests are proxied to Clang using HTTPS. Returns SOAP responses as JSON.

## Token

Authentication to the Clang Webservice API is done using a token (uuid). Contact the Support Desk of your Clang reseller to obtain a token. [Contact Leads to Loyals](https://www.leadstoloyals.nl/en/contact/) if you'd like to learn more about using Clang and this API.

Pass the token as an `uuid` header or a `_uuid` query parameter. The supplied token will be proxied to Clang with each request.

## Usage

- `GET /customers`
- `GET /customers/39515`
- `PUT /customers/39515`
  ```
  {
    "firstname": "Christiaan"
  }
  ```
- `POST /customers`
  ```
  {
    "firstname": "test",
    "emailAddress": "a@b.nl"
  }
  ```
- `DELETE /customers/39515`

The HTTP verb can be overriden using `_method` query parameter and you may send the parameters in the url as well.
- `GET /customers/39515?_method=PUT&firstname=Christiaan`
- `GET /customers?_method=POST&firstname=test&emailAddress=a@b.nl`
- `GET /customers/39515_method=DELETE`

(read more about [RESTful conventions](http://microformats.org/wiki/rest/urls))

- `GET /emails/26597`
- `POST /emails/26597/sendToCustomer`
  ```
  {
    "customerId": 328191
  }
  ```

The last example is a custom action on a single resource. That is only allowed using POST.

## Representation

Expect an array of objects when 200/201 or an object with a message property when an error occurred.

HTTP statusses as you would expect from a RESTful webservice remain intact.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 2019-01-22
### Changed
- Test of alle manual fields meegestuurd worden bij aanroep `email_sendToCustomer`

## [1.0.1] - 2018-12-26
### Changed
- Update README.md

## [1.0.0] - 2018-12-26
### Added
- POST en PUT methods parameters uit body gebruiken

### Changed
- Prettier, eslint en es6
- Losse routes en middleware gebruiken

### Removed
- Verouderde tests zijn komen te vervallen


test/test.js
//Get web.config variables into process.env
require('llips')().test.webconfig();
process.env['mocha-unfunk-style'] = 'css';

describe('Brands', require('./suites/brands'));
describe('Customers', require('./suites/customers'));


test/lib/base.js
const _ = require('lodash');

if (__dirname.match(/websites\\node\.leadstoloyals\.com\\/)) {
    exports.nodeUrl = 'https://node.leadstoloyals.com';
    exports.clangUrl = 'https://node.leadstoloyals.com/clang';
} else {
    exports.nodeUrl = 'https://node-dev.leadstoloyals.com';
    exports.clangUrl = 'https://node-dev.leadstoloyals.com/clang';
}

exports.anErrorCode = code => res => {
    if (res.body.profile) {
        res.body.should.have.property('error');
        if (res.body.error.length) {
            const myError = _.filter(res.body.error, { code: code });
            myError.should.be.an.Array.with.lengthOf(1);
            myError[0].should.have.property('code', code);
        } else {
            res.body.error.should.have.property('code', code);
        }
    } else if (code.length === 3) {
        //"213: Customer not found"
        res.body.should.have.property('code', 'NONE');
        res.body.should.have
            .property('message')
            .and.match(new RegExp('^' + code + ':.+'));
    } else {
        res.body.should.have.property('code', code);
    }
};


test/suites/brands.js
const config = require('../../config');
const lib = require('../../lib');
const llips = require('llips')(config, lib);
const base = require('../lib/base');

const supertest = require('supertest');
const requestClang = supertest(base.clangUrl);

const keys = Object.keys(base.brands);

module.exports = function() {
    keys.forEach(function(key) {
        it('Brand ' + key + ' should work', function(done) {
            this.timeout(10 * 1000); //10 seconde, ivm trage responstijd Clang
            requestClang
                .get('/customers/0')
                .set('uuid', base.brands[key])
                .expect('Content-Type', /json/)
                .expect(500)
                .expect(base.anErrorCode('213'))
                .end(llips.test.noIISnodeDevError(done));
        });
    });
};


server.js
const express = require('express');
const helmet = require('helmet');
const enforce = require('express-sslify');
const createError = require('http-errors');
const winston = require('winston');
const expressWinston = require('express-winston');
const clangRequest = require('./lib/clangRequest');
const prepare = require('./lib/prepare');

const port = process.env.PORT || 5000;

const app = express();
if (process.env.NODE_ENV === 'production') {
    app.use(helmet());
    app.use(enforce.HTTPS({ trustProtoHeader: true }));
}
app.use(express.json());

expressWinston.requestWhitelist.push('body', 'ip');
expressWinston.responseWhitelist.push('body');
app.use(
    expressWinston.logger({
        transports: [
            new winston.transports.Console({
                json: true,
                stringify: true,
                timestamp: true,
            }),
        ],
        expressFormat: true,
    })
);

app.get('/:object/:id?', prepare, (req, res, next) => {
    const { clangObjectName } = req.prepared;
    let clangMethodName;
    let args;

    const numKeys = Object.keys(req.query).length;
    if (req.params.id) {
        clangMethodName = 'getById';
        args = {
            ...req.query,
            [clangObjectName + 'Id']: req.params.id,
        };
    } else if (numKeys === 0) {
        clangMethodName = 'getAll';
        args = {};
    } else if (req.query['externalId'] && numKeys === 1) {
        clangMethodName = 'getByExternalId';
        args = {
            externalId: req.query['externalId'],
        };
    } else {
        clangMethodName = 'getByObject';
        args = {
            [clangObjectName]: { ...req.query },
        };
    }

    clangRequest(req, res, next, clangObjectName, clangMethodName, args);
});

app.post('/:object', prepare, (req, res, next) => {
    const { clangObjectName } = req.prepared;
    const clangMethodName = 'insert';
    const args = {
        [clangObjectName]: { ...req.query, ...req.body },
    };
    clangRequest(req, res, next, clangObjectName, clangMethodName, args);
});

app.post('/:object/:id/:customaction', prepare, (req, res, next) => {
    const { clangObjectName } = req.prepared;
    const { customaction: clangMethodName } = req.params;
    const args = {
        ...req.query,
        ...req.body,
        [clangObjectName + 'Id']: req.params.id,
    };
    clangRequest(req, res, next, clangObjectName, clangMethodName, args);
});

app.put('/:object/:id', prepare, (req, res, next) => {
    const { clangObjectName } = req.prepared;
    const clangMethodName = 'update';
    const args = {
        [clangObjectName]: {
            ...req.query,
            ...req.body,
            id: req.params.id,
        },
    };
    clangRequest(req, res, next, clangObjectName, clangMethodName, args);
});

app.delete('/:object/:id', prepare, (req, res, next) => {
    const { clangObjectName } = req.prepared;
    const clangMethodName = 'delete';
    const args = {
        [clangObjectName]: {
            id: req.params.id,
        },
    };
    clangRequest(req, res, next, clangObjectName, clangMethodName, args);
});

// catch 404 and forward to error handler
app.use((req, res, next) => {
    next(createError(404));
});
// eslint-disable-next-line no-unused-vars
app.use((err, req, res, next) => {
    res.status(err.status || 500);
    res.json({ error: err });
});

app.listen(port, () => console.log(`Listening on port ${port}`));


lib/index.js
exports.errors = {
  //start generic error codes
  '40101': 'Invalid api key or non provided.',
  '40102': 'uuid missing or invalid (add _uuid=... to your url)',
  '40401': 'Resource not found.',
  '50901': 'Invalid content-type {content-type}, must be {required-content-type}',
  '50902': 'Unknown fields: {fields}',
  '50903': 'A JSON object is required in the request body with these properties: {properties}.',
  '50904': 'All of these variables must be set to a value in the query string: {variables}.',
  '50905': 'Any of these variables must be set to a value in the query string: {variables}.',

  '50001': 'Clang api not created yet. Try again in a few seconds.',
  '50002': 'Clang resource not specified',
  '50003': 'Resource {resource} actually not available',
  '50004': 'Custom action invoked on unspecified resource (use /objects/123/customaction)',
  '50005': 'HTTP verb for this resource is not allowed',
  '50006': 'Method for this resource is not allowed'

};

lib/prepare.js
const createError = require('http-errors');

module.exports = (req, res, next) => {
    const uuid = req.headers.uuid || req.query._uuid;
    if (!uuid) {
        return next(
            createError(
                401,
                'uuid missing or invalid (add _uuid=... to your url)'
            )
        );
    }

    let clangObjectName = req.params.object;
    //remove a trailing s
    if (
        clangObjectName.match(/s$/) &&
        !clangObjectName.match(/sms|tatistics$/)
    ) {
        clangObjectName = clangObjectName.slice(0, -1);
    }

    const method = req.query._method || req.method; //HTTP VERB override through query paramater (override through http header would be better)

    delete req.query._method;
    delete req.query._uuid;

    req.prepared = {
        uuid,
        clangObjectName,
        method,
    };

    console.log('req.prepared', req.prepared);

    next();
};


lib/clangRequest.js
const Clang = require('clang');
const clang = new Clang({
    normalizeOptionFields: true,
});
const createError = require('http-errors');

const manualFieldRegex = /{{manual field='([^']*)'}}/g;
const getManualFields = string => {
    // https://stackoverflow.com/a/24795131/1385429
    const matches = [];
    string.replace(manualFieldRegex, (m, p1) => {
        matches.push(p1);
    });
    return matches;
};

const handleError = (err, req, next) => {
    let status = 500;
    if (
        err.Fault &&
        (err.Fault.faultcode == 213 ||
            err.Fault.faultstring.match(/not found/i))
    ) {
        status = 404;
    }
    if (err.Fault) {
        next(createError(status, err.Fault));
    } else {
        next(
            createError(status, {
                message: typeof err === 'string' ? err : err.message,
                method: req.prepared.method,
                params: req.params,
            })
        );
    }
};

module.exports = async (
    req,
    res,
    next,
    clangObjectName,
    clangMethodName,
    args
) => {
    const fullMethodName = `${clangObjectName}_${clangMethodName}`;
    const fullArgs = { ...args, uuid: req.prepared.uuid };

    // Check of er manual options missen, want email_sendToCustomer doet die check niet
    // Zie: https://github.com/leadstoloyals/aspnet-tickl-app/issues/464
    if (fullMethodName === 'email_sendToCustomer') {
        try {
            const result = await clang.request('email_getById', {
                uuid: fullArgs.uuid,
                emailId: fullArgs.emailId,
            });

            Object.entries(result).forEach(([key, value]) => {
                if (typeof value !== 'string') return;
                const manualFields = getManualFields(value);
                if (!manualFields.length) return;

                manualFields.forEach(manualField => {
                    if (fullArgs[manualField] == null) {
                        throw new Error(
                            `Manual field '${manualField}' required by '${key}' of email ${
                                fullArgs.emailId
                            } is missing in the arguments. Include '${manualField}' in the POSTed object to fix this.`
                        );
                    }
                });
            });
        } catch (err) {
            handleError(err, req, next);
            return; // stop requesting email_sendToCustomer
        }
    }

    try {
        const result = await clang.request(fullMethodName, fullArgs);
        res.status(200).json(result);
    } catch (err) {
        handleError(err, req, next);
    }
};


app.js
const express = require('express');
const helmet = require('helmet');
const enforce = require('express-sslify');
const winston = require('winston');
const expressWinston = require('express-winston');
const Clang = require('clang');

const clang = new Clang();

const port = process.env.PORT || 5000;

const app = express();
if (process.env.NODE_ENV === 'production') {
    app.use(helmet());
    app.use(enforce.HTTPS({ trustProtoHeader: true }));
}

expressWinston.requestWhitelist.push('body', 'ip');
expressWinston.responseWhitelist.push('body');
app.use(
    expressWinston.logger({
        transports: [
            new winston.transports.Console({
                json: true,
                stringify: true,
                timestamp: true,
            }),
        ],
        expressFormat: true,
    })
);

app.all('/:object/:id?/:customaction?', (req, res) => {
    const uuid = req.headers.uuid || req.query._uuid;
    if (
        !uuid ||
        !uuid.match(
            /^([0-9]-)?[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[89ab][a-f0-9]{3}-[a-f0-9]{12}$/i
        )
    ) {
        //Clang (probably) uses a version 4 UUIDs scheme relying only on random numbers.
        return res.status(401).send({
            message: 'uuid missing or invalid (add _uuid=... to your url)',
        });
    }

    let clangObjectName = req.params.object;
    //remove a trailing s
    if (
        clangObjectName.match(/s$/) &&
        !clangObjectName.match(/sms|tatistics$/)
    ) {
        clangObjectName = clangObjectName.slice(0, -1);
    }

    let clangMethodName;
    let args = {};
    const method = req.query._method || req.method; //HTTP VERB override through query paramater (override through http header would be better)

    delete req.query._method;
    delete req.query._uuid;
    const numKeys = Object.keys(req.query).length;

    switch (method) {
        case 'GET':
            if (req.params.id) {
                clangMethodName = 'getById';
                args = req.query;
                args[clangObjectName + 'Id'] = req.params.id;
            } else if (numKeys === 0) {
                clangMethodName = 'getAll';
            } else if (req.query['externalId'] && numKeys === 1) {
                clangMethodName = 'getByExternalId';
                args['externalId'] = req.query['externalId'];
            } else {
                clangMethodName = 'getByObject';
                args[clangObjectName] = req.query;
            }
            break;
        case 'POST':
            if (req.params.customaction) {
                clangMethodName = req.params.customaction; //override methodName with custom action like sendToCustomer (for POST only)
                if (req.params.id) {
                    args = req.query;
                    args[clangObjectName + 'Id'] = req.params.id;
                } else {
                    return res.status(500).send({
                        message:
                            'Custom action invoked on unspecified resource (use /objects/123/customaction)',
                    });
                }
            } else {
                clangMethodName = 'insert';
                args[clangObjectName] = req.query;
            }
            break;
        case 'PUT':
            clangMethodName = 'update';
            args[clangObjectName] = req.query;
            args[clangObjectName].id = req.params.id;
            break;
        case 'DELETE':
            clangMethodName = 'delete';
            args[clangObjectName].id = req.params.id;
            break;
        default:
            return res.status(405).send({
                message: 'HTTP verb for this resource is not allowed',
            });
    }

    args.uuid = uuid;

    clang.request(
        clangObjectName + '_' + clangMethodName,
        args,
        (err, result) => {
            if (err) {
                let status = 500;
                if (
                    err.Fault &&
                    (err.Fault.faultcode == 213 ||
                        err.Fault.faultstring.match(/not found/i))
                ) {
                    status = 404;
                }
                if (err.Fault) {
                    return res.status(status).send(err.Fault);
                }
                return res.status(status).send({
                    message: err.message,
                    method: method,
                    params: req.params,
                });
            }
            res.status(200).json(result);
        }
    );
});

const server = app.listen(port, () => {
    const host = server.address().address;

    console.log(`app listening at http://${host}:${port}`);
});
